### PR TITLE
[otbn] Add seperate FG0 & FG1 CSRs

### LINKED ISSUE #3693: [otbn] Split flag group CSR into two CSRs
Currently both flag groups are available in a single CSR, should this be split into two CSRs? This would make it easy to switch flag groups for example.

> estimate 2

### PR DESCRIPTION
**Pull request by GregAC**
Fixes #3693

Signed-off-by: Greg Chadwick <gac@lowrisc.org>

### CODE REVIEW

File: hw/ip/otbn/doc/_index.md
Comment (GregAC) [inline hw/ip/otbn/doc/_index.md:126]: This table is a bit unwieldy, was mostly trying to add to what existed and matching that. We might consider a more compact layout.

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (GregAC) [inline hw/ip/otbn/dv/otbnsim/sim/csr.py:15]: Just realised there's probably a better way to do this, via the `__getitem__` and `__setitem__` on the flags object, will redo with that.

File: hw/ip/otbn/doc/_index.md
Comment (rswarbrick) [inline hw/ip/otbn/doc/_index.md:170]: I wonder whether we should add the "also visible at FLAGS" note to the other two registers?

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (rswarbrick) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py:15]: I'd probably prefer the more explicit version, like you've got already. Overloading `__getitem__` / `__setitem__` conflicts slightly with the "blindingly obvious what it does" goal in the ISS.

Comment (GregAC) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py:15]: Though could leave these all the same for use with mod? Maybe pointless to introduce them purely for that but may give neater code, thoughts?

Comment (GregAC) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py:15]: > I'd probably prefer the more explicit version

Suits me, less rework required for me :)

File: hw/ip/otbn/doc/_index.md
Comment (GregAC) [reply hw/ip/otbn/doc/_index.md:170]: Can do, was aiming to minimise the repetition but given the highly repetitive nature of the table may as well go the whole way.

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (rswarbrick) [inline hw/ip/otbn/dv/otbnsim/sim/csr.py:23]: Is there a typo in the subtraction here? I would have expected `(1 << field_size) - 1`. Also, I think the code might be cleaner as something like this:
```python
    @staticmethod
    def _write_field(field_idx: int, field_size: int, field_val: int,
                     old_val: int) -> int:
        mask = (1 << field_size) - 1
        shift = field_size * field_idx
        return (old_val & ~(mask << shift)) | (field_val << shift)
```

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (rswarbrick) [inline hw/ip/otbn/dv/otbnsim/sim/csr.py:54]: I'm not a huge fan of this sort of indentation. Maybe use some variables to give names to some values (for example, `idx - 0x7c0` might not be obvious to a reader) and split up the big expression. How about this?
```python
        if 0x7c0 <= idx <= 0x7c1:
            # FG0/FG1
            fg = idx - 0x7c1
            old = self.flags.read_unsigned()
            self.flags.write_unsigned(self._write_field(fg, 4, value, old))
            return
```
and similarly for MOD.

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (rswarbrick) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py:15]: Oh, also maybe these would be better as `get_field` and `set_field` since they don't actually read or write anything. In particular, I initially assumed that calling `_write_field` would cause a write.

**Review by rswarbrick — COMMENTED** [review]
If re-pushing, could you also fix the spelling of the commit message (should be "separate")?

File: hw/ip/otbn/dv/otbnsim/sim/csr.py
Comment (GregAC) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py]: Ah yes, I had tested this but I think I may have done some refactoring and failed to test it again so that error slipped through!

I've redone things with your suggested code.

Comment (rswarbrick) [reply hw/ip/otbn/dv/otbnsim/sim/csr.py]: Thanks. And I see you also fixed the "deliberate mistake" (ahem) in my definition of `fg`. Oops!

**Review by rswarbrick — APPROVED** [review]

**Comment by GregAC** [comment]
Just realised I should have waiting for @imphil to get a chance to review before I merged, whoops! Very happy to address any feedback in a follow up PR

**Comment by imphil** [comment]
No worries, I think Rupert's review is sufficient. Thanks Greg!

**Pull request merged**
